File: neptune_py/discovery_service/skeleton.py
import uuid
import logging
import asyncio


class NeptuneService:
    def __init__(self, server, name=None):
        self.name = uuid.uuid4() if name is None else name
        self.server = server
        self._inited = False
        self._logic_task = None

    def init(self):
        """
        init service
        """
        if self._inited:
            return

        self._inited = True
        print(f"init service {self.name}")

    async def logic(self):
        """
        service logic
        """
        print(f"logic fun {self.name}")
        await asyncio.sleep(10)
        pass

    async def finished(self):
        """
        service finished
        """
        print(f"finished fun {self.name}")
        pass

    async def run(self):
        self.init()
        try:
            self._logic_task = asyncio.wait_for(self.logic(), None)
            await self._logic_task
        except Exception as e:
            if isinstance(e, asyncio.CancelledError):
                print(f'service {self.name} was cancelled')
            else:
                raise e
        finally:
            await self.finished()
            print(f'service {self.name} finished')

    def stop(self):
        if self._logic_task is not None and self._logic_task.done():
            self._logic_task.cancel()


class NeptuneServer:
    def __init__(self):
        self.services = set()

    def add_service(self, service: NeptuneService):
        self.services.add(service)

    def init():
        pass

    async def run(self):
        for service in self.services:
            service.init()

        ret = await asyncio.gather(
            *(service.run() for service in self.services),
            return_exceptions=True
        )
        print(ret)

async def main():
    server = NeptuneServer()
    server.add_service(NeptuneService(server))
    server.add_service(NeptuneService(server))
    await server.run()

asyncio.run(main())

File: neptune/src/skeleton/skeleton.go
package skeleton

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"
)

type NPServiceMeta struct {
	Service NeptuneService
	Ctx     context.Context
	Name    string
	Server  *NeptuneServerSkeleton
	Cancel  context.CancelFunc
}

type NeptuneServerSkeleton struct {
	services  []*NPServiceMeta
	waitGroup sync.WaitGroup
}

type NeptuneServiceStatus int

const (
	NeptuneServiceNew NeptuneServiceStatus = iota
	NeptuneServiceInited
	NeptuneServiceRunnning
	NeptuneServiceFinished
)

type NeptuneService interface {
	Init(*NeptuneServerSkeleton) error
	Logic(*NPServiceMeta) error
	Status() NeptuneServiceStatus
	Finish()
	Stop(*NPServiceMeta)
}

func (s *NeptuneServerSkeleton) AddService(name string, service NeptuneService) {
	ctx, cancel := context.WithCancel(context.Background())
	newService := &NPServiceMeta{
		Service: service,
		Name:    name,
		Ctx:     ctx,
		Server:  s,
		Cancel:  cancel,
	}

	s.services = append(s.services, newService)
}

func (s *NeptuneServerSkeleton) InitServices() {
	for _, service := range s.services {
		if service.Service.Status() != NeptuneServiceNew {
			log.Printf("cannot init service %s in status %v\n", service.Name, service.Service.Status())
			continue
		}
		if err := service.Service.Init(s); err != nil {
			log.Fatalf("error occurred during init %v\n", err)
		}
	}
}

func (s *NeptuneServerSkeleton) Run() {
	s.InitServices()
	log.Println("start running server")
	for _, service := range s.services {
		service := service // capture service
		s.waitGroup.Add(1)
		go func() {
			defer func() {
				service.Service.Finish()
				log.Printf("service %s has exited.\n", service.Name)
				s.waitGroup.Done()
			}()
			service.Service.Logic(service)
		}()
	}

	s.waitGroup.Wait()
	log.Printf("all services have finished, server will quit now.")
}

func (s *NeptuneServerSkeleton) FindService(name string) NeptuneService {
	for _, service := range s.services {
		if service.Name == name {
			return service.Service
		}
	}
	return nil
}

func (s *NeptuneServerSkeleton) CancelAll() {
	for _, service := range s.services {
		service.Service.Stop(service)
	}
}

type MyService int

func (s MyService) Init(ns *NeptuneServerSkeleton) error {
	fmt.Println("myservice init")
	return nil
}

func (s MyService) Logic(service *NPServiceMeta) error {
	fmt.Println("myservice logic")
	time.Sleep(time.Second * 3)
	return nil
}

func (s MyService) Status() NeptuneServiceStatus {
	return NeptuneServiceNew
}

func (s MyService) Finish() {
	fmt.Println("myservice finished")
}

func (s MyService) Stop(service *NPServiceMeta) {
	fmt.Println("stoping myservice")
	service.Cancel()
}
